docs/adr/0007-modelling-history.md

```diff
@@ -0,0 +1,272 @@
+
+- The ability to discard the draft and revert the current version in Content
+  Publisher back to the live version (including all images and attachments) to
+  be in sync with Publishing API;
+- The ability to have the details of the live version, most noticeably state
+  and URL so this can be presented to users;
+- Ability to re-present either the live or draft version to Publishing API
+  should there be any problems with publishing.
+
+And with regards to history:
+
+- To know who did what to a document and when - effectively an audit trail of
+  accountability;
+- Be able to compare two revisions of a document to show changes;
+- Have stored in our database a full history of a document, which may go back
+  further than the Publishing API (Whitehall has longer histories of some
+  published documents)
+
+A [MoSCoW](https://en.wikipedia.org/wiki/MoSCoW_method) breakdown of this
+functionality was defined as:
+
+### Must
+
+- Store full information on latest and live versions of a document
+- Consider how a revision of a document and aspects of it have state that may
+  change
+
+### Should
+
+- Have the full information for every edition that was published on GOV.UK
+- Record every change a publisher makes and have that information accessible
+- Consider states to be something that may change in a background job
+- Avoid relying on the TimelineEntry model as the sole answer to any historical
+  queries
+
+### Could
+
+- Link TimelineEntry to changes rather than storing the information twice
+- Enable a future feature to enable reverting to different points in time
+- Separate related data to different tables to reduce data usage
+
+### Won’t
+
+- Consider the sharing of data between translations, as per
+  [ADR-3](../adr/0003-initial-domain-modelling.md#decision)
+
+## Decision
+
+The proposed data model expands on the definitions of entities accordingly:
+
+**Document**: A record that represents all versions of a piece of content in a
+particular locale. It has many editions and many revisions. At any time it will
+have a current edition -  shown on Content Publisher index - and potentially a
+live edition which is currently on GOV.UK. The live and current edition can be
+the same. Document serves as a mutable entity that tracks when a Document was
+last edited and it is expected to be a joining point for document related data
+that is not associated with a particular editions.
+
+**Edition**: An edition is a numbered version of a document that has been, or is
+expected to be, published on GOV.UK. It is associated with a current revision
+and maintains an association with all revisions that have been associated with
+it. It has a current state (e.g. draft, submitted, published)
+and has knowledge of the past states it held. It is mutable and is expected
+as a place where any edition level database constraints can be placed.
+
+**Revision**: Represents the content of a document at a particular point in
```

Comment:
As discussed before monotonic is a new word for me :) so do let me know if I've misunderstood at all.

> * Should revision numbers be monotonic? I think this relates to an ambiguity on what happens to the history when a draft is discarded and a new one created, which might be worth clarifying up in the "features we'd like to enable".

Revision numbers should be sequential. We would still store a discarded draft and revisions created after the discarded draft would continue those numbers. As an example:

- I have a live edition with edition number 5 and which points to revision 16.
- I create a draft edition. It has edition number 6 and points to revision 17 (which is revision 16 with change note information reset).
- I amend that that draft edition it now points to revision 18.
- I discard the draft. The document now has a current edition of 5 which still points to revision 16.
- I create a draft edition. This draft has edition number 6 and points to revision 19 (which is revision 16 with change note information reset).

I'm not sure if that has explained things properly or not, so do let me know. I also picked a few tricky opinionated decisions in that scenario which I assume either make the system clearer or way more confusing. 🤞 

> * What about updating the live version of the document without discarding a draft? I notice this isn't in the desired features, so perhaps it will always be the current edition - perhaps it's better to say 'the same edition'.

That's not a supported feature. The model will be sufficient to support changing the state of a live edition while having a draft but I don't think (in the short term) we'd be support that through the UI. The changing of content is not intended to be supported for a live edition as there should only be one editable edition at a time.